google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java

```diff
@@ -811,6 +902,37 @@ public Batcher<RowMutationEntry, Void> newMutateRowsBatcher(
         MoreObjects.firstNonNull(ctx, clientContext.getDefaultCallContext()));
   }
 
+  /**
+   * Creates a {@link BatcherImpl} to handle {@link MutateRowsRequest.Entry} mutations. This is
+   * meant to be used for automatic batching with flow control.
+   *
+   * <ul>
+   *   <li>Uses {@link MutateRowsBatchingDescriptor} to spool the {@link RowMutationEntry} mutations
+   *       and send them out as {@link BulkMutation}.
+   *   <li>Uses {@link #bulkMutateRowsCallable()} to perform RPC.
+   *   <li>Batching thresholds can be configured from {@link
+   *       EnhancedBigtableStubSettings#bulkMutateRowsSettings()}.
+   *   <li>Process the response and schedule retries. At the end of each attempt, entries that have
+   *       been applied, are filtered from the next attempt. Also, any entries that failed with a
+   *       nontransient error, are filtered from the next attempt. This will continue until there
+   *       are no more entries or there are no more retry attempts left.
+   *   <li>Wrap batch failures in a {@link
+   *       com.google.cloud.bigtable.data.v2.models.MutateRowsException}.
+   *   <li>Split the responses using {@link MutateRowsBatchingDescriptor}.
+   * </ul>
+   */
+  public Batcher<RowMutationEntry, Void> newMutateRowsBatcher(
+      @Nonnull BulkMutation mutation, @Nullable GrpcCallContext ctx) {
```

Comment:
why do we need this API? the batcher is suppose to group RowMutationEntries into BulkMutations. Maybe instead we can do newMutateRowsBatcher(tableId, viewId, ctx)?

Reply:
I suppose we can do newMutateRowsBatcher(tableId, viewId, ctx).

But similar to the `exists()` and everything else, we didn't simply add a `viewId` argument to the existing function call but added a new `ExistsOption` argument. I think by doing this we avoid the potential growth of the arguments when adding of new features in the future, since we can simply add a new field in the options class rather than add a new API with more arguments.

Reply:
If you check the implementation of the current `newBulkReadRowsBatcher(
      String tableId, @Nullable Filters.Filter filter, @Nullable GrpcCallContext ctx)`[link](https://github.com/googleapis/java-bigtable/blob/main/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java#L1260-L1267):

```
public Batcher<ByteString, Row> newBulkReadRowsBatcher(
      String tableId, @Nullable Filters.Filter filter, @Nullable GrpcCallContext ctx) {
    Query query = Query.create(tableId);
    if (filter != null) {
      query.filter(filter);
    }
    return stub.newBulkReadRowsBatcher(query, ctx);
  }

```
It didn't use the `filter` argument directly. Instead, it converts the `filter` into a `Query` object and calls the `newBulkReadRowsBatcher()` on that query.

After this change, the `filter` argument will be wrapped in a `ReadRowsOptions` (which also wraps an authorized view id), and we still converts the `readRowsOptions` into a `Query` object first:

```
  public Batcher<ByteString, Row> newBulkReadRowsBatcher(
      String tableId, @Nonnull ReadRowsOptions readRowsOptions, @Nullable GrpcCallContext ctx) {
    Preconditions.checkNotNull(
        readRowsOptions, "Please use the newBulkReadRowsBatcher(tableId, ctx) API instead");
    Query query = Query.create(tableId, readRowsOptions);
    return stub.newBulkReadRowsBatcher(query, ctx);
  }
```

In terms of the mutation, in my opinion, the mutation-counterpart for a `Query` object is a `bulkMutation`, which is why I choosed to convert a `MutateRowOptions` into a `bulkMutation`, and then calls the `newMutateRowsBatcher()` on that `bulkMutation`. I think this way the bulk mutation implementation would be more consistent to the bulk readrows implementation.

Reply:
Also I wish to point out that even the original [newMutateRowsBatcher(
      @Nonnull String tableId, @Nullable GrpcCallContext ctx) implementation](https://github.com/googleapis/java-bigtable/blob/main/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java#L802-L812) actually created a `BulkMutation` object from the `tableId`.

I think in the end of the day, it is the `bulkMutateRowsCallable` requires a `BulkMutation` object, so we **have to** convert anything into a `BulkMutation` somewhere in the code.